### ISSUE TITLE
Why application state is validated only after commit?

### ISSUE THREAD

**Issue by skhoroshavin** [tendermint/tendermint#2494]
As far as I understand tendermint protocol transactions are applied to state and commited when enough pre-commits are received. Then AppHash returned from last commit is put into new proposed block, and other nodes check whether their last committed application state match. If application on some nodes incorrectly updates state (due to some bugs) this will be detected, but only after incorrect state is already committed. 

Why not apply transactions and calculate let's say uncommitted AppHash during proposal and pre-voting, and then commit it if it gets enough pre-commits, otherwise discard it? This way when less than 2/3 nodes agree block won't be committed and state won't be corrupted.

**Comment by ebuchman** [comment]
See some discussion in https://github.com/tendermint/tendermint/issues/2384

We plan to summarize this better in an ADR and will link to it from the documentation.

TL;DR: performance. Executing blocks first delays the consensus. It would also require some way to revert state over ABCI in case the executed block doesn't end up being committed.

We're considering a way to support executing blocks before committing them in a Tendermint 2.0, but this requires some research/design on using pipelining so we can execute blocks without delaying txs.

**Comment by ebuchman** [comment]
Note there's no state corruption in the current model - just the possibility of including "invalid" transactions. This will only happen if a proposer is malicious (this is guaranteed by the mempool, assuming the app is implemented correctly). It's the app's responsibility to punish proposers for including invalid txs. See the discussion in #2384 for more details.
